### DISCUSSION TITLE
Customizing Function Names in JFR for Flame Graph Remapping

### DISCUSSION THREAD

**Discussion by iambus** [async-profiler/async-profiler#987] [Q&A] [answered]
I have a specific requirement to transform the fully qualified Java function names in the flame graph for better readability.

I’m using Pyroscope, which leverages async-profiler, to profile JavaScript files compiled to Java classes by the Nashorn script engine. However, the flame graph shows the class names in a format that's difficult to interpret, and I'd like to remap them.

For example, the stack trace includes entries like:

```
jdk.nashorn.internal.scripts.Script$Recompilation$23$5872$\^eval\_.functionName(/org/example/ScriptName:205)
```

This appears in the flame graph as:

```
jdk.nashorn.internal.scripts.Script$Recompilation$23$5872$\^eval\_.functionName
```

Ideally, I would like the flame graph to display:

```
/org/example/ScriptName/functionName
```

Given that this is a special case, I'm open to implementing a custom solution. Could you provide any guidance on whether it's possible to modify the JFR data or function name processing to achieve this? If so, where should I begin?

**Comment by apangin** [comment]
Thank you for your interest in the project.
From the question, it's not quite clear what you expect from async-profiler in this regard, especially if you use it indirectly via third-party software. Is there a specific change that you'd like to make in async-profiler?

**Reply by iambus** [reply]
Thanks for the reply!

I was wondering if it’s possible for async-profiler to generate JFR files with modified stacktraces.

I’m not sure how async-profiler retrieves the stacktrace. If it can obtain the source file name for eachstack trace entry, maybe I could recompile a customized version of async-profiler and embed it into the pyroscope-java jar.

Any information would be appreciated!

**Reply by apangin** [reply, marked as answer]
Stacktrace is simply an array of method IDs. IDs are resolved to symbolic names when dumping results.
You may use JVM TI [GetSourceFileName](https://docs.oracle.com/en/java/javase/22/docs/specs/jvmti.html#GetSourceFileName) function to get a source file name for a class.

**Reply by iambus** [reply]
Thank you so much!
